Fix annotations for DataFrame.apply() including additional overloads

gandhis1 opened the pull request:
- [X] Tests added: Please use `assert_type()` to assert the type of any return value

In #401 the annotations for `apply` was changed. This introduced false negatives in my code base, because the particular mix of return types and arguments I was using was not supported. Here I am still not sure it is exhaustive but I think it's an improvement. In particular the `result_type` needs to be accounted for, as well as differentiating between a true list-like, `str`, and `Series` in the return value of the function.

@Dr-Irv @bashtage @danielroseman

gandhis1 commented on pandas-stubs/_typing.pyi:
```diff
@@ -155,6 +156,9 @@ AxisType: TypeAlias = Literal["columns", "index", 0, 1]
 ListLike = TypeVar("ListLike", Sequence, np.ndarray, "Series", "Index")
+ListLikeExceptSeriesAndStr = TypeVar(
+    "ListLikeExceptSeriesAndStr", MutableSequence, np.ndarray, tuple, "Index"
+)
```
I agree, the name is awful, but do we have a better way to exclude `str` from list-likes?

Dr-Irv replied:
I don't know of any

Dr-Irv reviewed (changes requested):
Can you add a test for `result_type="reduce"`, since you now differentiate between the possible values of `arg_type` ?

Otherwise looks good

gandhis1 commented on pandas-stubs/core/frame.pyi:
```diff
@@ -1086,36 +1089,178 @@ class DataFrame(NDFrame, OpsMixin):
+        **kwargs,
+    ) -> DataFrame: ...
+
+    # Add spacing between apply() overloads and remaining annotations
```
@Dr-Irv Do you mind looking over what I did with the overloads here? I believe it is correct to have to enumerate all the permutations of keyword only and default parameters, but not sure if this has unnecessary overloads or not.

I will note the reason why I separated out `axis=0` and `axis=1` is because the returned type in fact does change (when `result_type` = `None`).

I will note, MyPy seems to have trouble with one of the test cases where it thinks that `apply()` returns an `Any`. That makes no sense and I'm inclined to silence that. Pyright passes just fine.

Dr-Irv commented on pandas-stubs/core/frame.pyi:
```diff
@@ -1086,36 +1089,124 @@ class DataFrame(NDFrame, OpsMixin):
+        axis: AxisTypeColumn = ...,
+        result_type: Literal["broadcast"],
+        **kwargs,
+    ) -> DataFrame: ...
```
You could combine the above two overloads with the first one using `result_type: Literal["expand", "broadcast"] | None = ...`

I also think you want `axis: AxisTypeColumn` here without the `...` because these apply when the axis=1

Dr-Irv commented on tests/test_frame.py:
```diff
@@ -486,6 +486,182 @@ def gethead(s: pd.Series, y: int) -> pd.Series:
+    #         df.apply(returns_scalar, result_type="expand"), pd.DataFrame
+    #     ),
+    #     pd.DataFrame,
+    # )
```
When I run the above, it returns `Series` so it seems that you typed it wrong when testing.  Can you put it back using `pd.Series` ?

Dr-Irv replied:
It may make no sense, but I tried it and it worked.  In other words, it doesn't do nothing.  It does something!  So not sure what to do about that!

gandhis1 replied:
By nothing I mean that specifying `result_type="expand"` gives the same result as if you don't specify it (because it's not possible to expand a scalar).

The reason why it doesn't throw is because pandas has no way of knowing whether your function always returns a scalar or not, it has to assume that it's possible for some values to be scalars and some values to be list-like.

However, with annotations, we do know. Because a `Callable[[pd.Series], int]` can be assumed to always return an `int`.

I can certainly put it back, just felt that it might make sense to disallow it...

Dr-Irv replied:
I see your point, but we've taken the position that if the code works with pandas, we can't disallow it in the typing.  If someone is using `expand` with a `Scalar`, while it doesn't make sense, one could argue that our documentation doesn't say it is invalid.

gandhis1 replied:
Fair enough, I will uncomment these test cases and adjust the overloads as needed

Dr-Irv commented on tests/test_frame.py:
```diff
@@ -486,6 +486,182 @@ def gethead(s: pd.Series, y: int) -> pd.Series:
+    #         df.apply(returns_scalar, result_type="reduce"), pd.DataFrame
+    #     ),
+    #     pd.DataFrame,
+    # )
```
above returns `Series`, so can you fix that?

Dr-Irv commented on tests/test_frame.py:
```diff
@@ -486,6 +486,182 @@ def gethead(s: pd.Series, y: int) -> pd.Series:
+    #         df.apply(returns_series, result_type="reduce"), pd.Series
+    #     ),
+    #     pd.Series,
+    # )
```
The above should work, which means the typing is wrong.

Dr-Irv commented on tests/test_frame.py:
```diff
@@ -486,6 +486,182 @@ def gethead(s: pd.Series, y: int) -> pd.Series:
+    #         df.apply(returns_scalar, result_type="broadcast"), pd.DataFrame
+    #     ),
+    #     pd.DataFrame,
+    # )
```
The above should work if the typing is right.

Dr-Irv commented on tests/test_frame.py:
```diff
@@ -486,6 +486,182 @@ def gethead(s: pd.Series, y: int) -> pd.Series:
+    #     assert_type(
+    #         df.apply(returns_scalar, axis=1, result_type="expand"), pd.DataFrame
+    #     ),
+    #     pd.DataFrame,
```
should work but returns a `Series`.

Dr-Irv commented on tests/test_frame.py:
```diff
@@ -486,6 +486,182 @@ def gethead(s: pd.Series, y: int) -> pd.Series:
+    #         df.apply(returns_scalar, axis=1, result_type="reduce"), pd.DataFrame
+    #     ),
+    #     pd.DataFrame,
+    # )
```
above returns a `Series`, so fix the typing (and the test)

Dr-Irv commented on tests/test_frame.py:
```diff
@@ -486,6 +486,182 @@ def gethead(s: pd.Series, y: int) -> pd.Series:
+    #         df.apply(returns_series, axis=1, result_type="reduce"), pd.DataFrame
+    #     ),
+    #     pd.DataFrame,
+    # )
```
above should work

Dr-Irv commented on tests/test_frame.py:
```diff
@@ -486,6 +486,182 @@ def gethead(s: pd.Series, y: int) -> pd.Series:
+    #     assert_type(
+    #         df.apply(returns_scalar, axis=1, result_type="broadcast"), pd.DataFrame
+    #     ),
+    #     pd.DataFrame,
```
above should work

Dr-Irv reviewed (changes requested):
I Like the  additional tests, but all the ones you commented out do work, so we should accept them in typing.

gandhis1 commented:
I've updated the overloads to also handle the scalar value type when a `Series[T]` is returned. I've also updated the tests to include all cases which return a value.

Dr-Irv reviewed (approved):
thanks @gandhis1
